Assets/MixedRealityToolkit/_Core/Devices/UnityInput/OpenVR/UnityOpenVRDeviceManager.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using Microsoft.MixedReality.Toolkit.Internal.Definitions.Devices;
+using Microsoft.MixedReality.Toolkit.Internal.Definitions.Utilities;
+using System;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Internal.Devices.UnityInput.OpenVR
+{
+    public class UnityOpenVRDeviceManager : UnityDeviceManager
+    {
+        public UnityOpenVRDeviceManager(string name, uint priority) : base(name, priority) { }
+
+        #region Controller Utilities
+
+        /// <summary>
+        /// Retrieve the source controller from the Active Store, or create a new device and register it
+        /// </summary>
+        /// <returns>New or Existing Controller Input Source</returns>
+        protected override GenericUnityController GetOrAddController(string joystickName)
+        {
+            // If a device is already registered with the ID provided, just return it.
+            if (ActiveControllers.ContainsKey(joystickName))
+            {
+                var controller = ActiveControllers[joystickName];
+                Debug.Assert(controller != null);
+                return controller;
+            }
+
+            Handedness controllingHand;
+
+            if (joystickName.Contains("Left"))
+            {
+                controllingHand = Handedness.Left;
+            }
+            else if (joystickName.Contains("Right"))
+            {
+                controllingHand = Handedness.Right;
+            }
+            else
+            {
+                controllingHand = Handedness.None;
+            }
+
+            var currentControllerType = GetCurrentControllerType(joystickName);
+            Type controllerType;
+
+            switch (currentControllerType)
+            {
+                case SupportedControllerType.GenericOpenVR:
+                    controllerType = typeof(GenericUnityOpenVRController);
+                    break;
+                case SupportedControllerType.ViveWand:
+                    controllerType = typeof(UnityViveWandController);
+                    break;
+                case SupportedControllerType.ViveKnuckles:
+                    controllerType = typeof(UnityViveKnucklesController);
+                    break;
+                case SupportedControllerType.OculusTouch:
+                    controllerType = typeof(UnityOculusTouchController);
+                    break;
+                case SupportedControllerType.OculusRemote:
+                    controllerType = typeof(UnityOculusRemoteController);
+                    break;
+                case SupportedControllerType.WindowsMixedReality:
+                    controllerType = typeof(UnityWindowsMixedRealityUnityOpenVRController);
+                    break;
+                default:
+                    return null;
```

Comment:
Again, why is the default NULL?
If we don't recognise it in OpenVR, it should be the generic controllr NOT NULL

Reply:
because when null is returned, no input source is found and no input source detected is raised.

Reply:
If the controller is a generic ovr controller, when it'll already be assigned the type before we get here.

Reply:
See [Line 46](https://github.com/Microsoft/MixedRealityToolkit-Unity/pull/2568/files/528aa8eb742b881f4c5094b282c3ef854ea9595a#diff-42d2182336b6426682583fc7e7a29408R46)

In the `GetCurrentControllerType` method we should get a valid open vr controller type if the joystick name contains "OpenVR". If it comes back null, then it's not an open vr device.